Favourite controller

AhmedAbdEl-Qader opened the pull request.

Commit:
Added Favourite Controller (Endpoints, Handlers, Commands and Queries).
Added FavouriteDTOs

omar98alaa commented on Eskon.API/Controllers/FavouriteController.cs:
```diff
@@ -0,0 +1,64 @@
+
+        #region POST
+        [Authorize]
+        [HttpPost("Add")]
```
You need the `propertyId` param in route.
Change it to `HttpPost("{propertyId:guid}")`

omar98alaa commented on Eskon.API/Controllers/FavouriteController.cs:
```diff
@@ -0,0 +1,64 @@
+        #region GET
+        [Authorize]
+        [HttpGet("GetUserFavourites")]
+        public async Task<IActionResult> GetUserFavourites()
```
This should be paginated

omar98alaa commented on Eskon.API/Controllers/FavouriteController.cs:
```diff
@@ -0,0 +1,64 @@
+
+        #region DELETE
+        [Authorize]
+        [HttpDelete("Remove")]
```
`propertyId` param in route

omar98alaa commented on Eskon.Core/Features/FavouriteFeatures/Commands/Handler/FavouriteCommandHandler.cs:
```diff
@@ -0,0 +1,52 @@
+        }
+
+        public async Task<Response<string>> Handle(AddFavouriteCommand request, CancellationToken cancellationToken)
+        {
```
Before adding the new favourite, you should make checks including:

Check if property exists.
Check if a favourite already exists for the same user and property.

omar98alaa commented on Eskon.Core/Features/FavouriteFeatures/Commands/Handler/FavouriteCommandHandler.cs:
```diff
@@ -0,0 +1,52 @@
+            _mapper = mapper;
+        }
+
+        public async Task<Response<string>> Handle(AddFavouriteCommand request, CancellationToken cancellationToken)
```
You should return `Response<Favourite>` with the newly added entity.

omar98alaa commented on Eskon.API/Controllers/FavouriteController.cs:
```diff
@@ -0,0 +1,64 @@
+        #region DELETE
+        [Authorize]
+        [HttpDelete("Remove")]
+        public async Task<IActionResult> RemoveFromFavourites([FromRoute] Guid propertyId)
```
Delete should be with favoutiteId not propertyId

AhmedAbdEl-Qader replied:
There is no FavouriteId. If you mean the composite Primary Key userId + PropertyId
This is what is used here 
property Id from Route and UserId from Token

omar98alaa commented on Eskon.Core/Features/FavouriteFeatures/Commands/Handler/FavouriteCommandHandler.cs:
```diff
@@ -0,0 +1,52 @@
+        }
+        public async Task<Response<string>> Handle(RemoveFavouriteCommand request, CancellationToken cancellationToken)
+        {
+            var favourite = new Favourite
```
You should be getting favourite by Id then remove it

omar98alaa commented on Eskon.Core/Mapping/FavouriteMapping/FavouriteProfileMapping.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public partial class FavouriteProfileMapping : Profile
+    {
+        public FavouriteProfileMapping()
```
Register favourite mapping in `CoreDependencyInjection.cs`.
Make sure to follow file structure and naming.

omar98alaa commented on Eskon.Domian/DTOs/Favourite/FavouriteReadDTO.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Eskon.Domian.DTOs.Favourite
+{
+    public class FavouriteReadDTO
```
FavouriteReadDTO should be similar to PropertySummaryDTO

omar98alaa commented on Eskon.Domian/DTOs/Favourite/FavouriteRequestDTO.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Eskon.Domian.DTOs.Favourite
+{
+    public class FavouriteRequestDTO
```
FavouriteRequestDTO is implemented but you are not actually using it.
It is actually not needed since you get userId from token and propertyId from route.

AhmedAbdEl-Qader replied:
So, we don't need ReadDTO since we already have PropertyReadDTO, right?

omar98alaa reviewed (changes requested):
Needs fixing.
Please submit changes on multiple commits.

Commit:
Merge branch 'dev' of https://github.com/omar98alaa/Eskon into Favourite-Controller
Getting Latest Updates

Commit:
Added Checks before AddingNewFavourite.
AddFavourite returns FavouriteReadDTO instead of string as requested.
Modified RemoveCommandHandler as requested.

Commit:
Added GetFavouriteByID (using composed primary key) service and repo.
Added GetPaginatedFavourites and used it instead of generic.

Commit:
Allowed Pagination to GetUserFavourites

Commit:
Added mapping profile to CoreDependencyInjection.
Modifying DTOs.

Commit:
Fixed Paging in GetUserFavourites

omar98alaa committed:
Refactored implementation
